From e7f01c2dafb9cb9dfac491d6932abc268ed81b5a Mon Sep 17 00:00:00 2001
From: Kyle Brenneman <kbrenneman@nvidia.com>
Date: Wed, 2 Apr 2025 16:57:33 -0600
Subject: [PATCH 12/18] Update the X11 code to match the base library changes.

The big change for this is that the base library now uses an rwlock to
protect the surface list and the eglInitialize/eglTerminate state,
instead of a mutex.

Most EGL functions, most importantly eglSwapBuffers, only need to take a
read lock. That allows concurrent calls to eglSwapBuffers, while still
protecting against another thread trying to call eglDestroySurface or
eglTerminate at the same time.

As a result, we don't need a separate "in use" counter for EplDisplay,
EplSurface doesn't need to be refcounted, and we can get rid of all of
the temporary mutex unlock stuff in eglSwapBuffers.

Merged eplX11FreeWindow into eplX11DestroyWindow, since EplSurface isn't
refcounted anymore.

Changed to provide the SwapInterval and QuerySurface functions in
EplImplFuncs instead of providing the entire hook function.

Note that this brings in the base library's stub implementation of
EGL_EXT_buffer_age and EGL_KHR_partial_update, which will just return
zero for EGL_BUFFER_AGE_KHR. Adding a proper implementation of
EGL_BUFFER_AGE_KHR may come later, if one is needed.
---
 src/x11/x11-pixmap.c   |  10 +-
 src/x11/x11-platform.c |  68 +++---------
 src/x11/x11-platform.h |  10 +-
 src/x11/x11-window.c   | 246 +++++++++++------------------------------
 4 files changed, 87 insertions(+), 247 deletions(-)

diff --git a/src/x11/x11-pixmap.c b/src/x11/x11-pixmap.c
index e2d7dc4..f52175e 100644
--- a/src/x11/x11-pixmap.c
+++ b/src/x11/x11-pixmap.c
@@ -402,11 +402,12 @@ void eplX11DestroyPixmap(EplSurface *surf)
     }
 }
 
-static EGLBoolean CheckExistingPixmap(EplDisplay *pdpy, xcb_pixmap_t xpix)
+static EGLBoolean CheckExistingPixmap(EplDisplay *pdpy, xcb_pixmap_t xpix,
+        const struct glvnd_list *existing_surfaces)
 {
     EplSurface *psurf;
 
-    glvnd_list_for_each_entry(psurf, &pdpy->surface_list, entry)
+    glvnd_list_for_each_entry(psurf, existing_surfaces, entry)
     {
         if (psurf->type == EPL_SURFACE_TYPE_PIXMAP)
         {
@@ -424,7 +425,8 @@ static EGLBoolean CheckExistingPixmap(EplDisplay *pdpy, xcb_pixmap_t xpix)
 }
 
 EGLSurface eplX11CreatePixmapSurface(EplPlatformData *plat, EplDisplay *pdpy, EplSurface *surf,
-        EGLConfig config, void *native_surface, const EGLAttrib *attribs, EGLBoolean create_platform)
+        EGLConfig config, void *native_surface, const EGLAttrib *attribs, EGLBoolean create_platform,
+        const struct glvnd_list *existing_surfaces)
 {
     X11DisplayInstance *inst = pdpy->priv->inst;
     xcb_pixmap_t xpix = eplX11GetNativeXID(pdpy, native_surface, create_platform);
@@ -450,7 +452,7 @@ EGLSurface eplX11CreatePixmapSurface(EplPlatformData *plat, EplDisplay *pdpy, Ep
         eplSetError(plat, EGL_BAD_NATIVE_PIXMAP, "Invalid native pixmap %p\n", native_surface);
         return EGL_NO_SURFACE;
     }
-    if (!CheckExistingPixmap(pdpy, xpix))
+    if (!CheckExistingPixmap(pdpy, xpix, existing_surfaces))
     {
         return EGL_NO_SURFACE;
     }
diff --git a/src/x11/x11-platform.c b/src/x11/x11-platform.c
index a073f34..0b85eb5 100644
--- a/src/x11/x11-platform.c
+++ b/src/x11/x11-platform.c
@@ -76,17 +76,15 @@ static EGLBoolean eplX11GetPlatformDisplay(EplPlatformData *plat, EplDisplay *pd
         struct glvnd_list *existing_displays);
 static EGLBoolean eplX11InitializeDisplay(EplPlatformData *plat, EplDisplay *pdpy, EGLint *major, EGLint *minor);
 static void eplX11TerminateDisplay(EplPlatformData *plat, EplDisplay *pdpy);
-static void eplX11DestroySurface(EplDisplay *pdpy, EplSurface *surf);
-static void eplX11FreeSurface(EplDisplay *pdpy, EplSurface *surf);
+static void eplX11DestroySurface(EplDisplay *pdpy, EplSurface *surf,
+        const struct glvnd_list *existing_surfaces);
 static EGLBoolean eplX11WaitGL(EplDisplay *pdpy, EplSurface *psurf);
-static EGLBoolean eplX11HookQuerySurface(EGLDisplay edpy, EGLSurface esurf, EGLint attribute, EGLint *value);
+static EplQueryResult eplX11QuerySurface(EplDisplay *pdpy, EplSurface *psurf, EGLint attrib, EGLint *ret_value);
 
 static const EplHookFunc X11_HOOK_FUNCTIONS[] =
 {
     { "eglChooseConfig", eplX11HookChooseConfig },
     { "eglGetConfigAttrib", eplX11HookGetConfigAttrib },
-    { "eglQuerySurface", eplX11HookQuerySurface },
-    { "eglSwapInterval", eplX11SwapInterval },
 };
 static const int NUM_X11_HOOK_FUNCTIONS = sizeof(X11_HOOK_FUNCTIONS) / sizeof(X11_HOOK_FUNCTIONS[0]);
 
@@ -103,9 +101,10 @@ static const EplImplFuncs X11_IMPL_FUNCS =
     .CreateWindowSurface = eplX11CreateWindowSurface,
     .CreatePixmapSurface = eplX11CreatePixmapSurface,
     .DestroySurface = eplX11DestroySurface,
-    .FreeSurface = eplX11FreeSurface,
     .SwapBuffers = eplX11SwapBuffers,
     .WaitGL = eplX11WaitGL,
+    .SwapInterval = eplX11SwapInterval,
+    .QuerySurface = eplX11QuerySurface,
 };
 
 /**
@@ -1312,7 +1311,8 @@ static void eplX11TerminateDisplay(EplPlatformData *plat, EplDisplay *pdpy)
     pdpy->priv->inst = NULL;
 }
 
-static void eplX11DestroySurface(EplDisplay *pdpy, EplSurface *surf)
+static void eplX11DestroySurface(EplDisplay *pdpy, EplSurface *surf,
+        const struct glvnd_list *existing_surfaces)
 {
     if (surf->type == EPL_SURFACE_TYPE_WINDOW)
     {
@@ -1328,14 +1328,6 @@ static void eplX11DestroySurface(EplDisplay *pdpy, EplSurface *surf)
     }
 }
 
-static void eplX11FreeSurface(EplDisplay *pdpy, EplSurface *surf)
-{
-    if (surf->type == EPL_SURFACE_TYPE_WINDOW)
-    {
-        eplX11FreeWindow(surf);
-    }
-}
-
 static EGLBoolean eplX11WaitGL(EplDisplay *pdpy, EplSurface *psurf)
 {
     EGLBoolean ret = EGL_TRUE;
@@ -1551,54 +1543,22 @@ uint32_t eplX11GetNativeXID(EplDisplay *pdpy, void *native_surface, EGLBoolean c
     return xid;
 }
 
-EGLBoolean eplX11HookQuerySurface(EGLDisplay edpy, EGLSurface esurf, EGLint attribute, EGLint *value)
+EplQueryResult eplX11QuerySurface(EplDisplay *pdpy, EplSurface *psurf, EGLint attrib, EGLint *ret_value)
 {
-    EplDisplay *pdpy = NULL;
-    EplSurface *psurf = NULL;
-    EGLBoolean ret = EGL_FALSE;
-
-    pdpy = eplDisplayAcquire(edpy);
-    if (pdpy == NULL)
-    {
-        return EGL_FALSE;
-    }
-
-    if (value == NULL)
+    if (attrib == EGL_RENDER_BUFFER)
     {
-        eplSetError(pdpy->platform, EGL_BAD_ATTRIBUTE, "value pointer must not be NULL");
-        eplDisplayRelease(pdpy);
-        return EGL_FALSE;
-    }
-
-    psurf = eplSurfaceAcquire(pdpy, esurf);
-    if (psurf != NULL)
-    {
-        if (attribute == EGL_RENDER_BUFFER)
+        if (psurf->type == EPL_SURFACE_TYPE_WINDOW)
         {
-            if (psurf->type == EPL_SURFACE_TYPE_WINDOW)
-            {
-                *value = EGL_BACK_BUFFER;
-            }
-            else
-            {
-                *value = EGL_SINGLE_BUFFER;
-            }
-            ret = EGL_TRUE;
+            *ret_value = EGL_BACK_BUFFER;
         }
         else
         {
-            ret = pdpy->platform->priv->egl.QuerySurface(pdpy->internal_display,
-                    psurf->internal_surface, attribute, value);
+            *ret_value = EGL_SINGLE_BUFFER;
         }
-        eplSurfaceRelease(pdpy, psurf);
+        return EPL_QUERY_RESULT_SUCCESS;
     }
     else
     {
-        // If we don't recognize the EGLSurface, then just pass the call
-        // through to the driver.
-        ret = pdpy->platform->priv->egl.QuerySurface(pdpy->internal_display, esurf, attribute, value);
+        return EPL_QUERY_RESULT_UNKNOWN;
     }
-
-    eplDisplayRelease(pdpy);
-    return ret;
 }
diff --git a/src/x11/x11-platform.h b/src/x11/x11-platform.h
index bd0d163..3b6e791 100644
--- a/src/x11/x11-platform.h
+++ b/src/x11/x11-platform.h
@@ -397,10 +397,12 @@ EGLBoolean eplX11LoadEGLExternalPlatformCommon(int major, int minor,
         EGLint platform_enum);
 
 EGLSurface eplX11CreatePixmapSurface(EplPlatformData *plat, EplDisplay *pdpy, EplSurface *surf,
-        EGLConfig config, void *native_surface, const EGLAttrib *attribs, EGLBoolean create_platform);
+        EGLConfig config, void *native_surface, const EGLAttrib *attribs, EGLBoolean create_platform,
+        const struct glvnd_list *existing_surfaces);
 
 EGLSurface eplX11CreateWindowSurface(EplPlatformData *plat, EplDisplay *pdpy, EplSurface *surf,
-        EGLConfig config, void *native_surface, const EGLAttrib *attribs, EGLBoolean create_platform);
+        EGLConfig config, void *native_surface, const EGLAttrib *attribs, EGLBoolean create_platform,
+        const struct glvnd_list *existing_surfaces);
 
 EGLBoolean eplX11SwapBuffers(EplPlatformData *plat, EplDisplay *pdpy, EplSurface *surf,
         const EGLint *rects, EGLint n_rects);
@@ -452,12 +454,10 @@ EGLBoolean eplX11HookGetConfigAttrib(EGLDisplay edpy, EGLConfig config,
 
 void eplX11DestroyPixmap(EplSurface *surf);
 
-EGLBoolean eplX11SwapInterval(EGLDisplay edpy, EGLint interval);
+EGLBoolean eplX11SwapInterval(EplDisplay *pdpy, EplSurface *psurf, EGLint interval);
 
 void eplX11DestroyWindow(EplSurface *surf);
 
-void eplX11FreeWindow(EplSurface *surf);
-
 EGLBoolean eplX11WaitGLWindow(EplDisplay *pdpy, EplSurface *psurf);
 
 /**
diff --git a/src/x11/x11-window.c b/src/x11/x11-window.c
index 6040485..c7816fa 100644
--- a/src/x11/x11-window.c
+++ b/src/x11/x11-window.c
@@ -731,9 +731,45 @@ static EGLBoolean FindSupportedModifiers(X11DisplayInstance *inst,
     return EGL_TRUE;
 }
 
-void eplX11FreeWindow(EplSurface *surf)
+void eplX11DestroyWindow(EplSurface *surf)
 {
     X11Window *pwin = (X11Window *) surf->priv;
+    EGLSurface internalSurf;
+
+    assert(surf->type == EPL_SURFACE_TYPE_WINDOW);
+
+    /*
+     * Lock the surface and increment skip_update_callback. After that, if
+     * another thread tries to call the update callback after this, then
+     * the update callback won't try to call into the driver.
+     */
+    pthread_mutex_lock(&pwin->mutex);
+
+    pwin->skip_update_callback++;
+    internalSurf = surf->internal_surface;
+
+    pthread_mutex_unlock(&pwin->mutex);
+
+    /*
+     * We have to unlock the surface before we call the driver's
+     * eglDestroySurface.
+     *
+     * If another thread tries to call the window update callback for this
+     * surface, then it will be holding a mutex in the driver, and then the
+     * callback will try to lock the surface's mutex.
+     *
+     * If we had the surface locked here, then the driver's eglDestroySurface
+     * implementation would try to take the driver's mutex, which would lead to
+     * a deadlock.
+     */
+    if (internalSurf != EGL_NO_SURFACE)
+    {
+        pwin->inst->platform->egl.DestroySurface(pwin->inst->internal_display->edpy, internalSurf);
+    }
+
+    // After the driver's eglDestroySurface returns, no other threads will call
+    // the update or damage callbacks, so we can finish the rest of our
+    // cleanup.
 
     if (pwin == NULL)
     {
@@ -862,7 +898,7 @@ static EGLBoolean CheckReallocWindow(EplSurface *surf,
         *was_resized = EGL_FALSE;
     }
 
-    if (surf->deleted || pwin->native_destroyed)
+    if (pwin->native_destroyed)
     {
         return EGL_TRUE;
     }
@@ -952,7 +988,7 @@ static void PollForWindowEvents(EplSurface *surf)
 {
     X11Window *pwin = (X11Window *) surf->priv;
 
-    while (!pwin->native_destroyed && !surf->deleted)
+    while (!pwin->native_destroyed)
     {
         xcb_generic_event_t *xcbevt = xcb_poll_for_special_event(pwin->inst->conn, pwin->present_event);
         if (xcbevt == NULL)
@@ -970,20 +1006,6 @@ static void WindowUpdateCallback(void *param)
     EplSurface *surf = param;
     X11Window *pwin = (X11Window *) surf->priv;
 
-    /*
-     * Here, we lock the window mutex, but *not* the display mutex.
-     *
-     * We can't lock the display mutex, because we could run into a deadlock:
-     * - Thread A calls an EGL function in the platform library, which locks
-     *   the display. The platform library calls into the driver, which tries to
-     *   take the winsys lock.
-     * - Thread B calls a GL function. The driver takes the winsys lock, then
-     *   calls the window update callback. The window update callback tries to
-     *   take the display lock.
-     *
-     * We can safely access anything in X11DisplayInstance, because that never
-     * changes after it's initialized, and thus we don't need a mutex for it.
-     */
     pthread_mutex_lock(&pwin->mutex);
 
     if (pwin->skip_update_callback != 0)
@@ -1173,7 +1195,7 @@ static void WindowDamageCallback(void *param, int syncfd, unsigned int flags)
 
     // Check for any pending events first.
     PollForWindowEvents(surf);
-    if (pwin->native_destroyed || surf->deleted)
+    if (pwin->native_destroyed)
     {
         goto done;
     }
@@ -1249,11 +1271,12 @@ done:
     pthread_mutex_unlock(&pwin->mutex);
 }
 
-static EGLBoolean CheckExistingWindow(EplDisplay *pdpy, xcb_window_t xwin)
+static EGLBoolean CheckExistingWindow(EplDisplay *pdpy, xcb_window_t xwin,
+        const struct glvnd_list *existing_surfaces)
 {
     EplSurface *psurf;
 
-    glvnd_list_for_each_entry(psurf, &pdpy->surface_list, entry)
+    glvnd_list_for_each_entry(psurf, existing_surfaces, entry)
     {
         if (psurf->type == EPL_SURFACE_TYPE_WINDOW)
         {
@@ -1271,7 +1294,8 @@ static EGLBoolean CheckExistingWindow(EplDisplay *pdpy, xcb_window_t xwin)
 }
 
 EGLSurface eplX11CreateWindowSurface(EplPlatformData *plat, EplDisplay *pdpy, EplSurface *surf,
-        EGLConfig config, void *native_surface, const EGLAttrib *attribs, EGLBoolean create_platform)
+        EGLConfig config, void *native_surface, const EGLAttrib *attribs, EGLBoolean create_platform,
+        const struct glvnd_list *existing_surfaces)
 {
     X11DisplayInstance *inst = pdpy->priv->inst;
     xcb_window_t xwin = eplX11GetNativeXID(pdpy, native_surface, create_platform);
@@ -1299,7 +1323,7 @@ EGLSurface eplX11CreateWindowSurface(EplPlatformData *plat, EplDisplay *pdpy, Ep
         eplSetError(plat, EGL_BAD_NATIVE_WINDOW, "Invalid native window %p\n", native_surface);
         return EGL_NO_SURFACE;
     }
-    if (!CheckExistingWindow(pdpy, xwin))
+    if (!CheckExistingWindow(pdpy, xwin, existing_surfaces))
     {
         return EGL_NO_SURFACE;
     }
@@ -1458,7 +1482,7 @@ EGLSurface eplX11CreateWindowSurface(EplPlatformData *plat, EplDisplay *pdpy, Ep
 done:
     if (esurf == EGL_NO_SURFACE)
     {
-        eplX11FreeWindow(surf);
+        eplX11DestroyWindow(surf);
     }
     free(windowAttribReply);
     free(geomReply);
@@ -1469,67 +1493,14 @@ done:
     return esurf;
 }
 
-void eplX11DestroyWindow(EplSurface *surf)
-{
-    X11Window *pwin = (X11Window *) surf->priv;
-    EGLSurface internalSurf;
-
-    assert(surf->type == EPL_SURFACE_TYPE_WINDOW);
-
-    /*
-     * Lock the surface and increment skip_update_callback. After that, if
-     * another thread tries to call the update callback after this, then
-     * the update callback won't try to call into the driver.
-     */
-    pthread_mutex_lock(&pwin->mutex);
-
-    pwin->skip_update_callback++;
-    internalSurf = surf->internal_surface;
-
-    pthread_mutex_unlock(&pwin->mutex);
-
-    /*
-     * We have to unlock the surface before we call the driver's
-     * eglDestroySurface.
-     *
-     * If another thread tries to call the window update callback for this
-     * surface, then it will be holding a mutex in the driver, and then the
-     * callback will try to lock the surface's mutex.
-     *
-     * If we had the surface locked here, then the driver's eglDestroySurface
-     * implementation would try to take the driver's mutex, which would lead to
-     * a deadlock.
-     */
-    if (internalSurf != EGL_NO_SURFACE)
-    {
-        pwin->inst->platform->egl.DestroySurface(pwin->inst->internal_display->edpy, internalSurf);
-    }
-}
-
 /**
  * Waits for at least one Present event to arrive.
- *
- * This function will unlock the surface and the display while waiting, so the
- * caller must check the EplSurface::deleted flag to check whether another
- * thread destroyed the surface in the meantime.
  */
 static EGLBoolean WaitForWindowEvents(EplDisplay *pdpy, EplSurface *surf)
 {
     X11Window *pwin = (X11Window *) surf->priv;
     xcb_generic_event_t *xcbevt;
 
-    /*
-     * We don't want to block other threads while we wait, so we need to
-     * release our locks.
-     *
-     * The use counter in EplDisplay should prevent the display from
-     * getting terminated out from under us, and the refcount in EplSurface
-     * will ensure that the EplSurface struct itself sticks around.
-     *
-     * It's still possible that the surface will get destroyed by another
-     * thread in the meantime, though, so the caller needs to check for that.
-     */
-
     if (pwin->native_destroyed)
     {
         /*
@@ -1538,41 +1509,12 @@ static EGLBoolean WaitForWindowEvents(EplDisplay *pdpy, EplSurface *surf)
          * However, a new enough X server will send a PresentConfigureNotify
          * event with a special flag set to notify the client when that
          * happens.
-         *
-         * Note that even though we unlock the window's mutex before calling
-         * xcb_wait_for_special_event, we should still be safe from race
-         * conditions. This function is only called from eglSwapBuffers, which
-         * means that the window must be current. Thus, only one thread will
-         * ever be here for this window.
-         *
-         * There is still a race condition if the X window gets destroyed
-         * before the server receives the PresentPixmap request, but there's
-         * not much that we can do about that.
          */
         return EGL_TRUE;
     }
 
-    pthread_mutex_unlock(&pwin->mutex);
-    eplDisplayUnlock(pdpy);
-
     xcbevt = xcb_wait_for_special_event(pwin->inst->conn, pwin->present_event);
 
-    eplDisplayLock(pdpy);
-    pthread_mutex_lock(&pwin->mutex);
-
-    // Sanity check: If something called eglTerminate, then that should have
-    // destroyed the surface.
-    assert(pdpy->priv->inst == pwin->inst || surf->deleted);
-
-    if (surf->deleted)
-    {
-        /*
-         * Some other thread came along and called eglDestroySurface or
-         * eglTerminate.
-         */
-        return EGL_TRUE;
-    }
-
     if (xcbevt == NULL)
     {
         // Note that this only happens if the X11 connection gets killed, as
@@ -1797,26 +1739,10 @@ static int CheckBufferReleaseImplicit(EplDisplay *pdpy, EplSurface *surf,
         }
     }
 
-    // Release the locks while we wait, so that we don't block
-    // other threads.
-    pthread_mutex_unlock(&pwin->mutex);
-    eplDisplayUnlock(pdpy);
-
     ret = poll(fds, count, timeout_ms);
     err = errno;
 
-    eplDisplayLock(pdpy);
-    pthread_mutex_lock(&pwin->mutex);
-
-    if (surf->deleted)
-    {
-        /*
-         * Some other thread came along and called
-         * eglDestroySurface or eglTerminate.
-         */
-        return count;
-    }
-    else if (ret > 0)
+    if (ret > 0)
     {
         for (i=0; i<count; i++)
         {
@@ -1978,11 +1904,6 @@ static int CheckBufferReleaseExplicit(EplDisplay *pdpy, EplSurface *surf,
         timeout = 0;
     }
 
-    // Release the locks while we wait, so that we don't block
-    // other threads.
-    pthread_mutex_unlock(&pwin->mutex);
-    eplDisplayUnlock(pdpy);
-
     ret = pwin->inst->platform->priv->drm.SyncobjTimelineWait(
                 gbm_device_get_fd(pwin->inst->gbmdev),
                 handles, points, count, timeout,
@@ -1990,18 +1911,7 @@ static int CheckBufferReleaseExplicit(EplDisplay *pdpy, EplSurface *surf,
                 &first);
     err = errno;
 
-    eplDisplayLock(pdpy);
-    pthread_mutex_lock(&pwin->mutex);
-
-    if (surf->deleted)
-    {
-        /*
-         * Some other thread came along and called
-         * eglDestroySurface or eglTerminate.
-         */
-        return count;
-    }
-    else if (ret == 0)
+    if (ret == 0)
     {
         assert(first < count);
         if (WaitTimelinePoint(pwin->inst, &buffers[first]->timeline))
@@ -2085,7 +1995,7 @@ static X11ColorBuffer *GetFreeBuffer(EplDisplay *pdpy, EplSurface *surf,
         }
     }
 
-    while (!surf->deleted && !pwin->native_destroyed)
+    while (!pwin->native_destroyed)
     {
         X11ColorBuffer *buffer = NULL;
         int numBuffers = 0;
@@ -2180,12 +2090,7 @@ static X11ColorBuffer *GetFreeBuffer(EplDisplay *pdpy, EplSurface *surf,
 static EGLBoolean CheckWindowDeleted(EplSurface *surf, EGLBoolean *ret_success)
 {
     X11Window *pwin = (X11Window *) surf->priv;
-    if (surf->deleted)
-    {
-        *ret_success = EGL_TRUE;
-        return EGL_TRUE;
-    }
-    else if (pwin->native_destroyed)
+    if (pwin->native_destroyed)
     {
         *ret_success = EGL_FALSE;
         eplSetError(pwin->inst->platform, EGL_BAD_NATIVE_WINDOW, "The X11 window has been destroyed");
@@ -2370,50 +2275,23 @@ done:
     return ret;
 }
 
-EGLBoolean eplX11SwapInterval(EGLDisplay edpy, EGLint interval)
+EGLBoolean eplX11SwapInterval(EplDisplay *pdpy, EplSurface *psurf, EGLint interval)
 {
-    EplDisplay *pdpy = eplDisplayAcquire(edpy);
-    EGLSurface esurf;
-    EGLBoolean ret = EGL_FALSE;
-
-    if (pdpy == NULL)
+    if (psurf->type == EPL_SURFACE_TYPE_WINDOW)
     {
-        return EGL_FALSE;
-    }
+        X11Window *pwin = (X11Window *) psurf->priv;
 
-    esurf = pdpy->platform->egl.GetCurrentSurface(EGL_DRAW);
-    if (esurf != EGL_NO_SURFACE)
-    {
-        EplSurface *psurf = eplSurfaceAcquire(pdpy, esurf);
-        if (psurf != NULL)
-        {
-            if (psurf->type == EPL_SURFACE_TYPE_WINDOW)
-            {
-                X11Window *pwin = (X11Window *) psurf->priv;
-                pwin->swap_interval = interval;
-                if (pwin->swap_interval < 0)
-                {
-                    pwin->swap_interval = 0;
-                }
-            }
-            eplSurfaceRelease(pdpy, psurf);
-            ret = EGL_TRUE;
-        }
-        else
+        if (interval < 0)
         {
-            // If we don't recognize he current EGLSurface, then just pass the
-            // call through to the driver.
-            ret = pdpy->platform->priv->egl.SwapInterval(pdpy->internal_display, interval);
+            interval = 0;
         }
-    }
-    else
-    {
-        eplSetError(pdpy->platform, EGL_BAD_SURFACE, "eglSwapInterval called without a current EGLSurface");
-    }
 
-    eplDisplayRelease(pdpy);
+        pthread_mutex_lock(&pwin->mutex);
+        pwin->swap_interval = interval;
+        pthread_mutex_unlock(&pwin->mutex);
+    }
 
-    return ret;
+    return EGL_TRUE;
 }
 
 EGLBoolean eplX11WaitGLWindow(EplDisplay *pdpy, EplSurface *psurf)
@@ -2421,7 +2299,7 @@ EGLBoolean eplX11WaitGLWindow(EplDisplay *pdpy, EplSurface *psurf)
     X11Window *pwin = (X11Window *) psurf->priv;
 
     while ((pwin->last_present_serial - pwin->last_complete_serial) > 0
-            && !psurf->deleted && !pwin->native_destroyed)
+            && !pwin->native_destroyed)
     {
         if (!WaitForWindowEvents(pdpy, psurf))
         {
-- 
2.51.0

